### PR TITLE
add platt scaling for burns + fix for leace

### PR DESCRIPTION
**Pull request by lauritowal**
- Fixes platt scaling for Leace in CCS, where even during the training of the reporters the platt scaling parameters are multiplied and added to the raw_scores.

- Adds platt_scaling when using burns normalization

### CODE REVIEW

File: elk/training/train.py [hunk 1]
```diff
@@ -83,12 +83,9 @@ def apply_to_layer(
             reporter = CcsReporter(self.net, d, device=device, num_variants=v)
             train_loss = reporter.fit(first_train_h)
 
-            if not self.net.norm == "burns":
-                (_, v, k, _) = first_train_h.shape
-                reporter.platt_scale(
-                    to_one_hot(repeat(train_gt, "n -> (n v)", v=v), k).flatten(),
-                    rearrange(first_train_h, "n v k d -> (n v k) d"),
-                )
+            labels = to_one_hot(train_gt, k)
```
Comment (AlexTMallen) [inline elk/training/train.py:86]: this line doesn't do anything it seems

File: elk/training/train.py [hunk 2]
```diff
@@ -83,12 +83,9 @@ def apply_to_layer(
...
+            labels = repeat(train_gt, "n -> n v k", v=v, k=k)
```
Comment (AlexTMallen) [inline elk/training/train.py:87]: Why do we repeat the labels along the choices dimension? I think this means all dimensions of the one-hot prediction are expected to have the same value.

CCS only supports k=2, so we can just deal with that case after after asserting it true.

I think this line should be `labels = to_one_hot(repeat(train_gt, "n -> (n v)", v=v), 2).flatten()`

File: elk/training/ccs_reporter.py [hunk 3]
```diff
@@ -88,6 +88,8 @@ def __init__(
         num_variants: int = 1,
     ):
         super().__init__()
+        self._is_training = True
```
Comment (AlexTMallen) [inline elk/training/ccs_reporter.py:91]: if self._is_training is false during the call to platt_scale, then it won't work because the platt scaling parameter needs to be used for it to be updated in the backwards pass

**Review by AlexTMallen — CHANGES_REQUESTED** [review]

File: elk/training/ccs_reporter.py (hunk 3 above)
Comment (derpyplops) [inline elk/training/ccs_reporter.py:91]: 🟡 I think this approach basically works but perhaps the wrapper approach that the EigenFitter/Reporter uses is cleaner

File: elk/training/train.py (hunk 1 above)
Comment (lauritowal) [reply elk/training/train.py:86]: Oh nice catch! Thanks

File: elk/training/train.py (hunk 2 above)
Comment (lauritowal) [reply elk/training/train.py:87]: `labels = to_one_hot(repeat(train_gt, "n -> (n v)", v=v), 2).flatten()`
That wouldn't work, since we dimensions have to be the same for the labels and first_train_h, which is a 3D tensor in the case of CCS

**Review by derpyplops — CHANGES_REQUESTED** [review]

File: elk/training/ccs_reporter.py (hunk 3 above)
Comment (derpyplops) [reply elk/training/ccs_reporter.py:91]: Does it not?
```python
if self._is_training:
    return raw_scores
else:
    platt_scaled_scores = raw_scores.mul(self.scale).add(self.bias).squeeze(-1)
    return platt_scaled_scores
```

Anyway seems simple to test by comparing the results with and without the platt scaling params added

Comment (derpyplops) [reply elk/training/ccs_reporter.py:91]: In fact maybe we should just do that and make `CcsReporter.fit()` return a `Reporter`. Maybe need to have add a param that disables the eraser.

Comment (lauritowal) [reply elk/training/ccs_reporter.py]: Yeah, I removed the _is.training again. 

> In fact maybe we should just do that and make CcsReporter.fit() return a Reporter. Maybe need to have add a param that disables the eraser.

We can take a look at that. Could be maybe a new pull request or the same...

**Review by AlexTMallen — APPROVED** [review]
lgtm

File: elk/training/train.py [hunk 4]
```diff
@@ -82,13 +82,9 @@ def apply_to_layer(
...
             train_loss = reporter.fit(first_train_h)
-
-            if not self.net.norm == "burns":
-                (_, v, k, _) = first_train_h.shape
-                reporter.platt_scale(
-                    to_one_hot(repeat(train_gt, "n -> (n v)", v=v), k).flatten(),
-                    rearrange(first_train_h, "n v k d -> (n v k) d"),
-                )
+            labels = to_one_hot(train_gt, k)
+            labels = repeat(labels, "n k -> n v k", v=v)
```
Comment (derpyplops) [inline elk/training/train.py:86]: 🟢 nitpick: bugs like the one made here are the reason I dislike reassigning things to the same variable. Either the name should be changed:
`one_hotted_labels = to_one_hot(train_gt, k)`
or it should just be inlined
